File: 015. 三数之和.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        nums.sort()
        for i in range(len(nums)):
            if nums[i] > 0:
                break
            if i>0 and nums[i] == nums[i-1]:
                continue
            left = i + 1
            right = len(nums) - 1
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                if total < 0:
                    left += 1
                elif total > 0:
                    right -= 1
                else:
                    result.append([nums[i], nums[left], nums[right]])
                    # --- 核心去重逻辑 ---
                    # 只要左边下一个数和当前一样，就一直往右挪
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    # 只要右边下一个数和当前一样，就一直往左挪
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    # ------------------
                    left += 1
                    right -= 1
        return result
s = Solution()
print(s.threeSum([-1,0,1,2,-1,-4]))

File: 015. 三数之和.go
package main

import "sort"

func threeSum(nums []int) [][]int {
	// 先给数组排序，然后固定一个数，它的下一个数为左指针，最后一个数为右指针，然后求和。
	// 和小于0则右移左指针，大于0则左移右指针
	// 找到符合条件之后要继续找
	// 记得去重
	var res [][]int
	sort.Ints(nums)

	for i := range len(nums) - 1 {
		if nums[i] > 0 {
			break
		}
		if i != 0 && nums[i] == nums[i-1] {
			continue // 去重
		}

		l := i + 1
		r := len(nums) - 1
		for l < r {
			if l != i+1 && nums[l] == nums[l-1] {
				l++
				continue // 去重
			}
			if r != len(nums)-1 && nums[r] == nums[r+1] {
				r--
				continue // 去重
			}

			sum := nums[i] + nums[l] + nums[r]
			if sum < 0 {
				l++
			} else if sum > 0 {
				r--
			} else {
				res = append(res, []int{nums[i], nums[l], nums[r]})
				l++
				r--
			}
		}
	}
	return res
}
